python/objToJSON.c

```diff
@@ -342,7 +352,7 @@ static int SortedDict_iterNext(JSOBJ obj, JSONTypeContext *tc)
           goto error;
         }
         keyTmp = key;
-        key = PyUnicode_AsUTF8String(key);
+        key = PyUnicode_AsEncodedString(key, "utf-8", "surrogatepass");
```

Comment:
Yeah, agreed, it would be good to check the impact.

As of current CPython master:

* `PyUnicode_AsUTF8String` is simply a call to the internal `_PyUnicode_AsUTF8String` with a `NULL` error handler.
* `PyUnicode_AsEncodedString` runs `PyUnicode_Check`, `unicode_check_encoding_errors` (which does nothing in a normal release Python version without `-X dev`), and then compares the encoding string to `utf8` and variants. On match, it calls `_PyUnicode_AsUTF8String`. This was the shortest code path I found for UTF-8 encoding with an error handler.

There is a faster path in `PyUnicode_AsEncodedString` when `encoding` is `NULL`, which avoids the string comparison. However, this isn't explicitly documented. The method's description just says that

> *encoding* and *errors* have the same meaning as the parameters of the same name in the Unicode [encode()](https://docs.python.org/3/library/stdtypes.html#str.encode) method.

`str.encode`, of course, defaults to UTF-8. So `NULL` is practically the way to use the default `encoding` value, but I couldn't find any confirmation for that in the docs, so I'm hesitant to use it. This might be something to bring up upstream. If confirmed as intended/guaranteed behaviour, that'd basically cut it down to an unnecessary `PyUnicode_Check`.

Reply:
I looked into this again, and indeed, setting `encoding = NULL` is documented [here](https://docs.python.org/3/c-api/unicode.html#built-in-codecs):

> Setting encoding to `NULL` causes the default encoding to be used which is UTF-8.

Not sure how I missed that before...

So that's probably the best option and should reduce the impact of this change considerably as it gets rid of the encoding string normalisation and comparison on every call.